src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs

```diff
@@ -85,16 +85,14 @@ public TModel Train(TrainContext context)
         private void CheckInputSchema(SchemaShape inputSchema)
         {
             // Verify that all required input columns are present, and are of the same type.
-            var featureCol = inputSchema.FindColumn(FeatureColumn.Name);
-            if (featureCol == null)
+            if (!inputSchema.TryFindColumn(FeatureColumn.Name, out var featureCol))
                 throw Host.Except($"Feature column '{FeatureColumn.Name}' is not found");
             if (!FeatureColumn.IsCompatibleWith(featureCol))
                 throw Host.Except($"Feature column '{FeatureColumn.Name}' is not compatible");
```

Comment:
>not compatible" [](start = 77, length = 15)

this is not part of this PR, but I am wandering if the overall message will be clear enough: "not compatible with what, and how is it not compatible".

Reply:
Hi @sfilipi, indeed, this is one thing I noticed while changing the `IsCompatibleWith` method. You could imagine, perhaps a `CheckCompatibleWith` that takes an `ectx` and throws with more informative messages.

In fact during my testing before posting the PR, I would have found this quite useful -- I had in my first draft of the code a handful of small bugs here and there, as we may expect. But when running the test I got the programmatic equivalent to "no, guess again," even though the chain of boolean tests is fairly exhaustive and one could easily insert the throws in there. (I do not think there is much of any conceivable reason to call the method unless you're planning to throw immediately. I mean, it's not like the single `bool` is actually actionable information.)

Yet, I feel like this code is still in flux -- I saw a review comment from @Zruty0 that suggests that this method in particular will be reworked. But maybe he wants to start cogitating on it so I will mention it.

---
In reply to: [215756284](https://github.com/dotnet/machinelearning/pull/845#discussion_r215756284) [](ancestors = 215756284)

Reply:
Incidentally if you're looking for excellent ways to reward me for giving such a thoughtful answer, a re-review followed by a possibly approval might be helpful, since AFAIK the only thing standing in the way is that goofy built break...

---
In reply to: [215785227](https://github.com/dotnet/machinelearning/pull/845#discussion_r215785227) [](ancestors = 215785227,215756284)

Reply:
:)

---
In reply to: [215785972](https://github.com/dotnet/machinelearning/pull/845#discussion_r215785972) [](ancestors = 215785972,215785227,215756284)